crates/ruff_linter/src/message/sarif.rs

```diff
@@ -0,0 +1,195 @@
+        }
+        self.applied_rules = applied_rules;
+        self
+    }
+}
+
+impl Emitter for SarifEmitter<'_> {
+    fn emit(
+        &mut self,
+        writer: &mut dyn Write,
+        messages: &[Message],
+        _context: &EmitterContext,
+    ) -> anyhow::Result<()> {
+        let results = messages
+            .iter()
+            .map(|message| SarifResult::from_message(message))
+            .collect::<Vec<_>>();
+
+        let output = json!({
+            "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
+            "version": "2.1.0",
+            "runs": [{
+                "tool": {
+                    "driver": {
+                        "name": "ruff",
+                        "informationUri": "https://github.com/astral-sh/ruff",
+                        "rules": self.applied_rules,
+                        "version": VERSION.to_string(),
+                    }
+                },
+                "results": results,
+            }],
+        });
+        serde_json::to_writer_pretty(writer, &output)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SarifRule<'a> {
+    name: &'a str,
+    code: String,
+    linter: &'a str,
+    summary: &'a str,
+    explanation: Option<&'a str>,
+    url: Option<String>,
+}
+
+impl<'a> SarifRule<'a> {
+    fn from_rule(rule: Rule) -> Self {
+        let code = rule.noqa_code().to_string();
+        let (linter, _) = Linter::parse_code(&code).unwrap();
+        Self {
+            name: rule.to_owned().into(),
+            code,
+            linter: linter.name(),
+            summary: rule.message_formats()[0],
+            explanation: rule.explanation(),
+            url: rule.url(),
+        }
+    }
+}
+
+impl Serialize for SarifRule<'_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        json!({
+            "id": self.code,
+            "shortDescription": {
+                "text": self.summary,
+            },
+            "fullDescription": {
+                "text": self.explanation,
+            },
+            "helpUri": self.url,
+            "properties": {
+                "id": self.code,
+                "kind": self.linter,
+                "name": self.name,
+                "problem.severity": "error".to_string(),
+            },
+        })
+        .serialize(serializer)
+    }
+}
+
+struct SarifResult {
+    rule: Rule,
+    level: String,
+    message: String,
+    uri: String,
+    start_line: OneIndexed,
+    start_column: OneIndexed,
+}
+
+impl SarifResult {
+    fn from_message(message: &Message) -> Self {
+        let start_location = message.compute_start_location();
+        Self {
+            rule: message.kind.rule(),
+            level: "error".to_string(),
+            message: message.kind.name.to_owned(),
+            uri: Url::from_file_path(message.filename()).unwrap().to_string(),
```

Comment:
If `filename` is relative here, would this convert it to absolute? I do think `filename` will always be absolute here... we pass around absolute paths, and expect callers to call `relativize_path` when displaying user-facing relative paths... But we may want to be defensive (call `.canonicalize()` or similar).

Reply:
This is what I was referring to in the PR body.  It'll error if the path is relative, which is why I changed the path in the snapshot files.  I'll update to safely handle relative paths, should be straight forward.

Reply:
👍 I did see that, but wasn't sure how `Url::from_file_path` would behave if the path wasn't absolute.

Reply:
yeah, `normalize_path` seems like the right call here, as `.canonicalize()` will fail if the file doesn't actually exist (a la in test).